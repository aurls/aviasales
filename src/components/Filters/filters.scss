@import '../../assets/styles/styleConstants';

.filters {
  box-shadow: 0 2px 8px rgba(0, 0, 0, .1);
  border-radius: 5px;
  padding: 20px;
  background: white;
}

.filters__title {
  font-size: 12px;
  line-height: 12px;
  font-weight: 600;
  letter-spacing: .5px;
  text-transform: uppercase;
}

.filter {
  display: flex;
  align-items: center;
  height: 40px;
  margin: 0 -20px 0 -20px;
  padding: 0 20px 0 20px;
  cursor: pointer;
  font-size: 13px;
  line-height: 20px;
  user-select: none;

  &:first-child {
    margin-top: 10px;
  }

  &:last-child {
    margin-bottom: -10px;
  }

  &:hover {
    background: $lightPrimaryColor;
    transition: all $standardTransition;
  }
}

.filter__checkbox {
  position: relative;
  width: 20px;
  height: 20px;
  margin: 0 10px 0 0;
  border: 1px solid #9abbce;
  border-radius: 2px;
  padding: 0;
  appearance: none;
  cursor: inherit;
  outline: none;

  &:checked {
    border-color: #2196f3;

    &::before {
      content: "";
      position: absolute;
      top: 5px;
      left: 3px;
      display: block;
      width: 12px;
      height: 8px;
      background: url("./check.png");
      background-size: 100%;
    }
  }
}
