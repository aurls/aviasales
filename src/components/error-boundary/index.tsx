import React from 'react';
import ErrorMessage from '../error-message';
import './error-boundary.scss';

type Props = {
  children: React.ReactNode
}

type State = {
  hasReactError: boolean
}

class ErrorBoundary extends React.Component<Props, State> {
  state: State = {
    hasReactError: false
  };

  componentDidCatch () {
    this.setState({
      hasReactError: true
    });
  }

  render () {
    const { hasReactError } = this.state;
    if (hasReactError) {
      return (
        <div className="error-boundary">
          <ErrorMessage />
        </div>);
    }
    return this.props.children;
  }
}

export default ErrorBoundary;
