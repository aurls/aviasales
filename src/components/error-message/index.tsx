import React from 'react';
import './error-message.scss';

const ErrorMessage: React.FC = () => {
  return (
    <section className="error-message">
      <h2 className="error-message__title">
        Ой, у&nbsp;нас что-то случилось!
      </h2>
      <p className="error-message__text">
        Мы уже знаем о&nbsp;проблеме и&nbsp;кинули все силы на&nbsp;её&nbsp;устранение.
        Попробуйте поискать через 10&nbsp;минут
      </p>
    </section>
  );
};

export default ErrorMessage;
