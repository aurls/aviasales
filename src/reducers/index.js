import { combineReducers } from 'redux';

import ticketsReducer from './ticketsReducer';
import filterReducer from './filterReducer';
import sortingReducer from './sortingReducer';
import appearanceReducer from './appearanceReducer';

const rootReducer = combineReducers({
  tickets: ticketsReducer,
  appliedFilters: filterReducer,
  appliedSorting: sortingReducer,
  appearance: appearanceReducer
});

export default rootReducer;
