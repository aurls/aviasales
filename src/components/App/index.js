import React, { Fragment } from 'react';

import Header from '../Header';
import Filters from '../Filters';
import Sorting from '../Sorting';
import Tickets from '../Tickets';
import './app.scss';

class App extends React.Component {
  render () {
    return (
      <Fragment>
        <Header />
        <div className="app">
          <aside className="app__sidebar">
            <Filters />
          </aside>
          <main className="app__main">
            <Sorting />
            <Tickets />
          </main>
        </div>
      </Fragment>
    );
  }
}

export default App;
