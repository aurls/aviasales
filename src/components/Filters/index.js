import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as filters from '../../constants/filters';
import { filters as filtersActions } from '../../actions';
import './filters.scss';

const Filters = ({ appliedFilters, setFilter }) => {
  const isFilterChecked = (filter) => {
    return appliedFilters.includes(filters.ALL.type) || appliedFilters.includes(filter.type);
  };

  const onChangeFilter = (event) => {
    setFilter(event.target.value);
  };

  return (
    <div className="filters">
      <h2 className="filters__title">Количество пересадок</h2>
      <ul>
        {
          Object.values(filters).map((filter) =>
            <li key={filter.type}>
              <label className="filter"
                htmlFor={`filter-${filter.type}`}>
                <input className="filter__checkbox"
                  id={`filter-${filter.type}`}
                  type="checkbox"
                  name="filter"
                  value={filter.type}
                  checked={isFilterChecked(filter)}
                  onChange={onChangeFilter} />
                {filter.name}
              </label>
            </li>)
        }
      </ul>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    appliedFilters: state.appliedFilters
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setFilter: bindActionCreators(filtersActions.setFilter, dispatch)
  };
};

Filters.propTypes = {
  appliedFilters: PropTypes.array.isRequired,
  setFilter: PropTypes.func.isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Filters);
