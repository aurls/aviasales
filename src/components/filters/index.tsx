import React from 'react';
import Types from '../../types';
import actions from '../../store/actions';
import filters from '../../services/filters';
import './filters.scss';

type Props = {
  currentFilters: Types.Filter[]
  setFilter: typeof actions.setFilter
}

const Filters: React.FC<Props> = (props: Props) => {
  const {
    currentFilters,
    setFilter
  } = props;

  const isFilterChecked = (filter: Types.Filter) => {
    return currentFilters.includes(filter) || currentFilters.includes(filters.ALL);
  };

  const onFilterClick = (filter: Types.Filter) => {
    setFilter(filter);
  };

  return (
    <div className="filters">
      <h2 className="filters__title">Количество пересадок</h2>
      <ul>
        {
          Object.values(filters).map((filter) =>
            <li key={filter.type}>
              <label className="filter">
                <input className="filter__checkbox"
                  type="checkbox"
                  name="filter"
                  value={filter.type}
                  checked={isFilterChecked(filter)}
                  onChange={() => onFilterClick(filter)} />
                {filter.title}
              </label>
            </li>)
        }
      </ul>
    </div>
  );
};

export default React.memo(Filters);
