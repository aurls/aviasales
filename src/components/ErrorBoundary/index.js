import React from 'react';
import PropTypes from 'prop-types';

import ErrorMessage from '../ErrorMessage';
import './error-boundary.scss';

class ErrorBoundary extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      hasReactError: false
    };
  }

  componentDidCatch () {
    this.setState({
      hasReactError: true
    });
  }

  render () {
    const { hasReactError } = this.state;
    if (hasReactError) {
      return (
        <div className="error-boundary">
          <ErrorMessage />
        </div>);
    }
    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.element.isRequired
};

export default ErrorBoundary;
