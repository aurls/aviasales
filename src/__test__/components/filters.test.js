import React from 'react';

import { Filters } from '../../components/Filters';
import { initialState as appliedFilters } from '../../reducers/filterReducer';

describe('Filters component', () => {
  const setFilter = jest.fn();
  let component;

  beforeEach(() => {
    component = shallow(
      <Filters
        appliedFilters={appliedFilters}
        setFilter={setFilter} />
    );
  });

  it('Default state', () => {
    expect(component).toMatchSnapshot();
  });
});
