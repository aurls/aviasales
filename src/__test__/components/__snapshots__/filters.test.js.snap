// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component Default state 1`] = `
<div
  className="filters"
>
  <h2
    className="filters__title"
  >
    Количество пересадок
  </h2>
  <ul>
    <li
      key="ALL"
    >
      <label
        className="filter"
        htmlFor="filter-ALL"
      >
        <input
          checked={true}
          className="filter__checkbox"
          id="filter-ALL"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="ALL"
        />
        Все
      </label>
    </li>
    <li
      key="NO_TRANSFERS"
    >
      <label
        className="filter"
        htmlFor="filter-NO_TRANSFERS"
      >
        <input
          checked={true}
          className="filter__checkbox"
          id="filter-NO_TRANSFERS"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="NO_TRANSFERS"
        />
        Без пересадок
      </label>
    </li>
    <li
      key="ONE_TRANSFER"
    >
      <label
        className="filter"
        htmlFor="filter-ONE_TRANSFER"
      >
        <input
          checked={true}
          className="filter__checkbox"
          id="filter-ONE_TRANSFER"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="ONE_TRANSFER"
        />
        1 пересадка
      </label>
    </li>
    <li
      key="TWO_TRANSFERS"
    >
      <label
        className="filter"
        htmlFor="filter-TWO_TRANSFERS"
      >
        <input
          checked={true}
          className="filter__checkbox"
          id="filter-TWO_TRANSFERS"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="TWO_TRANSFERS"
        />
        2 пересадки
      </label>
    </li>
    <li
      key="THREE_TRANSFERS"
    >
      <label
        className="filter"
        htmlFor="filter-THREE_TRANSFERS"
      >
        <input
          checked={true}
          className="filter__checkbox"
          id="filter-THREE_TRANSFERS"
          name="filter"
          onChange={[Function]}
          type="checkbox"
          value="THREE_TRANSFERS"
        />
        3 пересадки
      </label>
    </li>
  </ul>
</div>
`;
